### ISSUE TITLE
[question] mini dsl

### ISSUE THREAD

**Issue by emicklei** [antonmedv/expr#98]
hi,

I started writing a REPL tool that uses `expr` for evaluating each entry.

Each entry expression is using/calling Go structs and functions but I want to allow a different syntax for calling functions. Currently, functions need to be exposed (Upcase) and when no arguments are expected, a function call require open/closing brackets to write.

For example, I would like to evaluate this
```
x.foo.bar
```
as if it was written using Go syntax like this
```
x.Foo().Bar()
```
pre: method Foo exists on x and method Bar exists on the result type of Foo.

I see a few approaches to work on (in a fork?):
- add callback function to allow the resolving of unknown fields
- make the program return Error objects with information such that an entry can be rewritten 
- some aliasing feature, e.g. MapFieldToMethod(new(X), "foo", "Foo")

Thoughts?

**Comment by antonmedv** [comment]
Hi,

This is a really strange syntax for calling func: `x.foo.bar`. But if you want to do it anyway I think the best way is to do some patching of AST. Here is an example:
 
https://github.com/antonmedv/expr/blob/master/compiler/patcher.go

You can find more examples of patching here: 

https://github.com/antonmedv/expr/blob/master/optimizer/in_range.go

Use visitor to traverse AST, find property accessor node, replace with the corresponding call method.

**Comment by antonmedv** [comment]
Also you can use Reflection to determine correct method names and so on.

**Comment by emicklei** [comment]
Thank you for responding to my question.

I started working with your suggestion to create a patcher.
In order change a step in the Compiler process, I need to decompose the expr.Compile.
First, I created a new function in my package to compile a parser.Tree with this signature

```
func Compile(tree *parser.Tree, ops ...expr.Option) (*vm.Program, error) 
```
However, because the definition uses the internal/conf package, this does not work.
I prefer not to fork your package so do you see a way for me to make it work? What about adding this to `expr`:

```
func BuildConfig(ops ...expr.Option) (*conf.Config, error)
```
and then I can have
```
func Compile(tree *parser.Tree, conf *conf.Config) (*vm.Program, error) 
```
hmm , still would not work because I cannot import `expr/internal/conf`

**Comment by antonmedv** [comment]
Yes, I see. This internal package was a bad idea. I think to remove it at all. Make conf a top-level package.

**Comment by antonmedv** [comment]
@emicklei I've creaded small PR with new functinality what actially may help. 
https://github.com/antonmedv/expr/pull/99

**Comment by antonmedv** [comment]
Released in v1.7.0. Now it should be easy to implement your dsl.
